expirationd.lua

```diff
@@ -258,9 +258,30 @@ local function worker_loop(task)
     -- detach worker from the guardian and attach it to sched fiber
     fiber.name(string.format("worker of %q", task.name), { truncate = true })
 
+    -- https://www.tarantool.io/en/doc/latest/reference/configuration/#confval-read_only
+    local space = box.space[task.space_id]
+    local is_ro_writable = space.temporary or space.is_local
+    local is_skipped = false
     while true do
-        local replication = box.cfg.replication_source or box.cfg.replication
-        if replication == nil or #replication == 0 or task.force then
+        if not box.info.ro or is_ro_writable or task.force then
+            if is_skipped then
+                -- Wait for maximum upstream lag * 2 if we became a master.
+                -- It helps to don't hit a conflict due to parallel processing
+                -- of remaining transactions from the old master and from the
+                -- new master.
+                local max_lag = 0
+                for i = 1, table.maxn(box.info.replication) do
+                    local r = box.info.replication[i]
```

Comment:
Probably it's better to call box.info once at loop start.
Because box.info() call produces really huge table and it increases pressure on GC.

Reply:
I don't think that RO -> RW switching to be too frequent. This is a rare event.

Also we always have a delay in the loop:
```lua
 314         -- Full scan iteration is complete, yield                               
 315         fiber.sleep(task.full_scan_delay) 
 ```
 
 I don't think it will cause any problems.

Reply:
Unfortunately, we need to check `box.info.ro` here too. But iterations are not too frequent here (full index processing + delay).

Reply:
In pseudo-code the loop looks like:
```Lua
while true do
    if box.info.ro or other then
         get max lag from box.info.replication
    end
    full index scan with tuple processing and delays
    delay
end
```
I think an iteration will take a long time, so that it won't be a problem to check `box.info`.

Reply:
I guess this comment is about calling box.info twice in the same iteration (for `.ro` and for `.replication`).

Reply:
Hmmm, I didn't think from this point of view. I got it. Thank you!
```lua
--- a/expirationd.lua
+++ b/expirationd.lua
@@ -263,15 +263,16 @@ local function worker_loop(task)
     local is_ro_writable = space.temporary or space.is_local
     local is_skipped = false
     while true do
-        if not box.info.ro or is_ro_writable or task.force then
+        local box_info = box.info
+        if not box_info.ro or is_ro_writable or task.force then
             if is_skipped then
                 -- Wait for maximum upstream lag * 2 if we became a master.
                 -- It helps to don't hit a conflict due to parallel processing
                 -- of remaining transactions from the old master and from the
                 -- new master.
                 local max_lag = 0
-                for i = 1, table.maxn(box.info.replication) do
-                    local r = box.info.replication[i]
+                for i = 1, table.maxn(box_info.replication) do
+                    local r = box_info.replication[i]
                     if r and r.upstream and max_lag < r.upstream.lag then
                         max_lag = r.upstream.lag
                     end
```